File: Task_1/benchmark.py
import time
import statistics
import csv
import os
import psutil
from matrix import multiplicar_matrices as multiply_matrices

def estimate_memory_mb(n):
    return 3 * n * n * 8 / (1024 * 1024)

def benchmark(n, runs):
    times = []
    process = psutil.Process(os.getpid())

    for i in range(runs):
        memory_before = process.memory_info().rss / (1024 * 1024)
        start = time.perf_counter()
        multiply_matrices(n)
        end = time.perf_counter()
        memory_after = process.memory_info().rss / (1024 * 1024)
        duration = end - start
        used_memory = memory_after - memory_before
        times.append((duration, used_memory))
        print(f"Running {i+1}/{runs}: {duration:.4f} seconds, Memory used: {used_memory:.2f} MB")

    durations = [t[0] for t in times]
    memory_usages = [t[1] for t in times]

    mean = statistics.mean(durations)
    stdev = statistics.stdev(durations)
    minimum = min(durations)
    maximum = max(durations)
    real_memory = statistics.mean(memory_usages)
    theoretical_memory = estimate_memory_mb(n)

    print("\nResults:")
    print(f"Matrix size: {n}x{n}")
    print(f"Number of runs: {runs}")
    print(f"Average time: {mean:.4f} s")
    print(f"Standard deviation: {stdev:.4f} s")
    print(f"Best time: {minimum:.4f} s")
    print(f"Worst time: {maximum:.4f} s")
    print(f"Average real memory used: {real_memory:.2f} MB")
    print(f"Theoretical memory usage: {theoretical_memory:.2f} MB")

    save_result("python", n, runs, mean, stdev, minimum, maximum, real_memory, theoretical_memory)

def save_result(language, n, runs, mean, stdev, minimum, maximum, real_memory, theoretical_memory):
    folder = "results"
    os.makedirs(folder, exist_ok=True)
    file = os.path.join(folder, "benchmark_results.csv")
    exists = os.path.isfile(file)

    with open(file, "a", newline="") as f:
        writer = csv.writer(f)
        if not exists:
            writer.writerow([
                "language", "size", "runs", "mean (s)", "stdev",
                "best", "worst", "real memory (MB)", "theoretical memory (MB)"
            ])
        writer.writerow([
            language, n, runs,
            f"{mean:.4f}", f"{stdev:.4f}", f"{minimum:.4f}", f"{maximum:.4f}",
            f"{real_memory:.2f}", f"{theoretical_memory:.2f}"
        ])

if __name__ == "__main__":
    n = int(input("Enter matrix size: "))
    runs = int(input("Enter number of runs: "))
    benchmark(n, runs)

File: Task_1/Benchmark.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
#include <chrono>
#include <algorithm>
#include <filesystem>
#include <sys/resource.h>
#include "Matrix.hpp"

double estimate_memory_MB(int n) {
    return 3.0 * n * n * 8 / (1024.0 * 1024.0);
}

double get_memory_MB() {
    struct rusage usage;
    getrusage(RUSAGE_SELF, &usage);
    return usage.ru_maxrss / 1024.0;
}

double mean(const std::vector<double>& v) {
    double sum = 0.0;
    for (double x : v) sum += x;
    return sum / v.size();
}

double stdev(const std::vector<double>& v, double m) {
    if (v.size() <= 1) return 0.0;
    double sum = 0.0;
    for (double x : v) sum += std::pow(x - m, 2);
    return std::sqrt(sum / (v.size() - 1));
}

int main() {
    int n, runs;
    std::cout << "Enter matrix size: ";
    std::cin >> n;
    std::cout << "Enter number of runs: ";
    std::cin >> runs;

    std::vector<double> times;
    std::vector<double> memory_usages;

    for (int i = 0; i < runs; i++) {
        double memory_before = get_memory_MB();
        auto start = std::chrono::high_resolution_clock::now();
        multiply_matrices(n);
        auto end = std::chrono::high_resolution_clock::now();
        double memory_after = get_memory_MB();

        std::chrono::duration<double> duration = end - start;
        double used_memory = std::max(0.0, memory_after - memory_before);

        times.push_back(duration.count());
        memory_usages.push_back(used_memory);

        std::cout << "Running " << (i + 1) << "/" << runs
                  << ": " << duration.count()
                  << " seconds, Memory used: " << used_memory << " MB\n";
    }

    double mean_time = mean(times);
    double sd_time = stdev(times, mean_time);
    double min_time = *std::min_element(times.begin(), times.end());
    double max_time = *std::max_element(times.begin(), times.end());
    double real_memory = mean(memory_usages);
    double theoretical_memory = estimate_memory_MB(n);

    std::cout << "\nResults:\n";
    std::cout << "Matrix size: " << n << "x" << n << "\n";
    std::cout << "Number of runs: " << runs << "\n";
    std::cout << "Average time: " << mean_time << " s\n";
    std::cout << "Standard deviation: " << sd_time << " s\n";
    std::cout << "Best time: " << min_time << " s\n";
    std::cout << "Worst time: " << max_time << " s\n";
    std::cout << "Average real memory used: " << real_memory << " MB\n";
    std::cout << "Theoretical memory usage: " << theoretical_memory << " MB\n";

    std::filesystem::create_directory("results");
    std::ofstream file("results/benchmark_results.csv", std::ios::app);
    if (file.tellp() == 0)
        file << "language,size,runs,mean (s),stdev,best,worst,real memory (MB),theoretical memory (MB)\n";

    file << "cpp," << n << "," << runs << ","
         << mean_time << "," << sd_time << ","
         << min_time << "," << max_time << ","
         << real_memory << "," << theoretical_memory << "\n";

    file.close();


    return 0;
}
